File: anp/direct_e2ee/prekey_manager.py
"""Prekey helpers for direct_e2ee.

[INPUT]: Local DID metadata, Ed25519 signing material, signed prekey storage,
remote RPC callbacks, prekey bundle objects, and issuer DID documents.
[OUTPUT]: Signed prekeys, strict Appendix-B prekey bundles, published RPC
payloads, and validation errors for invalid remote bundles.
[POS]: This module is the direct-E2EE adapter that maps the shared Appendix-B
object proof core onto ``prekey_bundle`` semantics.

[PROTOCOL]:
1. Treat ``owner_did`` as the issuer DID for every prekey bundle proof.
2. Require Appendix-B object proof validation for remote bundle acceptance.
3. Keep bundle business validation separate from proof validation.
"""

from __future__ import annotations

import base64
import time
from typing import Any, Callable, Dict, Optional

from cryptography.hazmat.primitives.asymmetric import ed25519
from cryptography.hazmat.primitives.asymmetric.x25519 import (
    X25519PrivateKey,
    X25519PublicKey,
)

from anp.authentication.did_wba import validate_did_document_binding
from anp.e2e_encryption_hpke.key_pair import extract_x25519_public_key_from_did_document
from anp.proof import generate_object_proof, verify_object_proof

from .errors import DirectE2eeError
from .models import MTI_DIRECT_E2EE_SUITE, PrekeyBundle, SignedPrekey
from .store import FileSignedPrekeyStore

RPCClient = Callable[[str, Dict[str, Any]], Dict[str, Any]]


def _b64u_encode(data: bytes) -> str:
    return base64.urlsafe_b64encode(data).decode("ascii").rstrip("=")


class PrekeyManager:
    """Manage signed prekeys and bundle publication."""

    def __init__(
        self,
        local_did: str,
        static_key_agreement_id: str,
        signing_private_key: ed25519.Ed25519PrivateKey,
        signing_verification_method: str,
        signed_prekey_store: FileSignedPrekeyStore,
        rpc_client: Optional[RPCClient] = None,
    ) -> None:
        self._local_did = local_did
        self._static_key_agreement_id = static_key_agreement_id
        self._signing_private_key = signing_private_key
        self._signing_verification_method = signing_verification_method
        self._signed_prekey_store = signed_prekey_store
        self._rpc_client = rpc_client

    def generate_signed_prekey(
        self,
        key_id: str,
        expires_at: str,
    ) -> tuple[X25519PrivateKey, SignedPrekey]:
        private_key = X25519PrivateKey.generate()
        public_key = private_key.public_key().public_bytes_raw()
        metadata = SignedPrekey(
            key_id=key_id,
            public_key_b64u=_b64u_encode(public_key),
            expires_at=expires_at,
        )
        self._signed_prekey_store.save_signed_prekey(key_id, private_key, metadata)
        return private_key, metadata

    def build_prekey_bundle(
        self,
        signed_prekey: SignedPrekey,
        bundle_id: Optional[str] = None,
        created: Optional[str] = None,
    ) -> PrekeyBundle:
        bundle_id = bundle_id or f"spk-{int(time.time())}-{signed_prekey.key_id}"
        unsigned_bundle = {
            "bundle_id": bundle_id,
            "owner_did": self._local_did,
            "suite": MTI_DIRECT_E2EE_SUITE,
            "static_key_agreement_id": self._static_key_agreement_id,
            "signed_prekey": signed_prekey.to_dict(),
        }
        signed_bundle = generate_object_proof(
            unsigned_bundle,
            self._signing_private_key,
            self._signing_verification_method,
            issuer_did=self._local_did,
            created=created,
        )
        return PrekeyBundle(
            bundle_id=bundle_id,
            owner_did=self._local_did,
            suite=MTI_DIRECT_E2EE_SUITE,
            static_key_agreement_id=self._static_key_agreement_id,
            signed_prekey=signed_prekey,
            proof=signed_bundle["proof"],
        )

    def publish_prekey_bundle(self, bundle: PrekeyBundle) -> Dict[str, Any]:
        if self._rpc_client is None:
            raise DirectE2eeError("RPC client is not configured", "rpc_unavailable")
        return self._rpc_client(
            "direct.e2ee.publish_prekey_bundle",
            {
                "meta": {
                    "anp_version": "1.0",
                    "profile": "anp.direct.e2ee.v1",
                    "security_profile": "transport-protected",
                    "sender_did": self._local_did,
                    "operation_id": f"op-publish-{bundle.bundle_id}",
                },
                "body": {
                    "prekey_bundle": bundle.to_dict(),
                },
            },
        )

    def ensure_fresh_prekey_bundle(self) -> PrekeyBundle:
        latest = self._signed_prekey_store.load_latest_signed_prekey()
        if latest is None:
            _, signed_prekey = self.generate_signed_prekey(
                "spk-initial",
                "2030-01-01T00:00:00Z",
            )
            bundle = self.build_prekey_bundle(signed_prekey)
            if self._rpc_client is not None:
                self.publish_prekey_bundle(bundle)
            return bundle
        _, metadata = latest
        bundle = self.build_prekey_bundle(metadata)
        if self._rpc_client is not None:
            self.publish_prekey_bundle(bundle)
        return bundle

    @staticmethod
    def verify_prekey_bundle(bundle: PrekeyBundle, did_document: Dict[str, Any]) -> None:
        if did_document.get("id") != bundle.owner_did:
            raise DirectE2eeError(
                "owner_did must match the issuer DID document",
                "bundle_invalid",
            )
        if did_document.get("id", "").startswith("did:wba:") and not validate_did_document_binding(
            did_document,
            verify_proof=False,
        ):
            raise DirectE2eeError(
                "owner DID document binding validation failed",
                "bundle_invalid",
            )
        key_agreement_ids = {
            item if isinstance(item, str) else item.get("id")
            for item in did_document.get("keyAgreement", [])
        }
        if bundle.static_key_agreement_id not in key_agreement_ids:
            raise DirectE2eeError(
                "static_key_agreement_id must appear in DID document keyAgreement",
                "bundle_invalid",
            )
        try:
            verify_object_proof(
                bundle.to_dict(),
                issuer_did=bundle.owner_did,
                issuer_did_document=did_document,
            )
        except Exception as exc:
            raise DirectE2eeError("bundle proof verification failed", "bundle_invalid") from exc

    @staticmethod
    def extract_recipient_static_key(
        did_document: Dict[str, Any],
        key_id: str,
    ) -> tuple[X25519PublicKey, str]:
        return extract_x25519_public_key_from_did_document(did_document, key_id)

File: golang/direct_e2ee/prekey_manager.go
package directe2ee

import (
	"crypto/ecdh"
	"fmt"
	"sort"
	"time"

	anp "github.com/agent-network-protocol/anp/golang"
)

// RPCClient invokes the remote message-service RPC boundary.
type RPCClient func(method string, params map[string]any) (map[string]any, error)

// PrekeyManager manages signed prekeys and bundle publication.
type PrekeyManager struct {
	localDID                  string
	localServiceDID           string
	staticKeyAgreementID      string
	signingPrivateKey         anp.PrivateKeyMaterial
	signingVerificationMethod string
	signedPrekeyStore         SignedPrekeyStore
	oneTimePrekeyStore        OneTimePrekeyStore
	rpcClient                 RPCClient
}

const defaultOneTimePrekeyBatchSize = 16

// NewPrekeyManager creates a direct E2EE prekey manager.
func NewPrekeyManager(localDID string, localServiceDID string, staticKeyAgreementID string, signingPrivateKey anp.PrivateKeyMaterial, signingVerificationMethod string, signedPrekeyStore SignedPrekeyStore, rpcClient RPCClient, oneTimePrekeyStores ...OneTimePrekeyStore) *PrekeyManager {
	var oneTimePrekeyStore OneTimePrekeyStore
	if len(oneTimePrekeyStores) > 0 {
		oneTimePrekeyStore = oneTimePrekeyStores[0]
	}
	return &PrekeyManager{localDID: localDID, localServiceDID: localServiceDID, staticKeyAgreementID: staticKeyAgreementID, signingPrivateKey: signingPrivateKey, signingVerificationMethod: signingVerificationMethod, signedPrekeyStore: signedPrekeyStore, oneTimePrekeyStore: oneTimePrekeyStore, rpcClient: rpcClient}
}

// GenerateSignedPrekey generates, stores, and returns a new signed prekey.
func (m *PrekeyManager) GenerateSignedPrekey(keyID string, expiresAt string) (*ecdh.PrivateKey, SignedPrekey, error) {
	if expiresAt == "" {
		expiresAt = defaultSignedPrekeyExpiry()
	}
	privateMaterial, publicMaterial, _, err := anp.GenerateKeyPairPEM(anp.KeyTypeX25519)
	if err != nil {
		return nil, SignedPrekey{}, err
	}
	privateKey, err := ecdh.X25519().NewPrivateKey(privateMaterial.Bytes)
	if err != nil {
		return nil, SignedPrekey{}, err
	}
	metadata := SignedPrekey{KeyID: keyID, PublicKeyB64U: anp.EncodeBase64URL(publicMaterial.Bytes), ExpiresAt: expiresAt}
	if err := m.signedPrekeyStore.SaveSignedPrekey(keyID, privateMaterial, metadata); err != nil {
		return nil, SignedPrekey{}, err
	}
	return privateKey, metadata, nil
}

// GenerateOneTimePrekey generates, stores, and returns a new OPK.
func (m *PrekeyManager) GenerateOneTimePrekey(keyID string) (*ecdh.PrivateKey, OneTimePrekey, error) {
	if m.oneTimePrekeyStore == nil {
		return nil, OneTimePrekey{}, &Error{Code: "store_unavailable", Message: "one-time prekey store is not configured"}
	}
	privateMaterial, publicMaterial, _, err := anp.GenerateKeyPairPEM(anp.KeyTypeX25519)
	if err != nil {
		return nil, OneTimePrekey{}, err
	}
	privateKey, err := ecdh.X25519().NewPrivateKey(privateMaterial.Bytes)
	if err != nil {
		return nil, OneTimePrekey{}, err
	}
	metadata := OneTimePrekey{KeyID: keyID, PublicKeyB64U: anp.EncodeBase64URL(publicMaterial.Bytes)}
	if err := m.oneTimePrekeyStore.SaveOneTimePrekey(keyID, privateMaterial, metadata); err != nil {
		return nil, OneTimePrekey{}, err
	}
	return privateKey, metadata, nil
}

// BuildPrekeyBundle signs a prekey bundle.
func (m *PrekeyManager) BuildPrekeyBundle(signedPrekey SignedPrekey, bundleID string, created string) (PrekeyBundle, error) {
	if bundleID == "" {
		bundleID = fmt.Sprintf("spk-%d-%s", time.Now().Unix(), signedPrekey.KeyID)
	}
	return BuildPrekeyBundle(bundleID, m.localDID, m.staticKeyAgreementID, signedPrekey, m.signingPrivateKey, m.signingVerificationMethod, created)
}

// PublishPrekeyBundle publishes a prekey bundle over the RPC boundary.
func (m *PrekeyManager) PublishPrekeyBundle(bundle PrekeyBundle) (map[string]any, error) {
	if m.rpcClient == nil {
		return nil, &Error{Code: "rpc_unavailable", Message: "RPC client is not configured"}
	}
	body := map[string]any{"prekey_bundle": bundleToMap(bundle)}
	oneTimePrekeys, err := m.listOneTimePrekeys()
	if err != nil {
		return nil, err
	}
	if len(oneTimePrekeys) > 0 {
		body["one_time_prekeys"] = oneTimePrekeysToAny(oneTimePrekeys)
	}
	return m.rpcClient("direct.e2ee.publish_prekey_bundle", map[string]any{"meta": map[string]any{"anp_version": "1.0", "profile": "anp.direct.e2ee.v1", "security_profile": "transport-protected", "sender_did": m.localDID, "target": map[string]any{"kind": "service", "did": m.localServiceDID}, "operation_id": "op-publish-" + bundle.BundleID}, "body": body})
}

// EnsureFreshPrekeyBundle returns the latest prekey bundle or creates one.
func (m *PrekeyManager) EnsureFreshPrekeyBundle() (PrekeyBundle, error) {
	if _, err := m.EnsureFreshOneTimePrekeys(defaultOneTimePrekeyBatchSize); err != nil {
		return PrekeyBundle{}, err
	}
	_, metadata, ok, err := m.signedPrekeyStore.LoadLatestSignedPrekey()
	if err != nil {
		return PrekeyBundle{}, err
	}
	if !ok {
		_, signedPrekey, err := m.GenerateSignedPrekey("spk-initial", "2030-01-01T00:00:00Z")
		if err != nil {
			return PrekeyBundle{}, err
		}
		bundle, err := m.BuildPrekeyBundle(signedPrekey, "", "")
		if err != nil {
			return PrekeyBundle{}, err
		}
		if m.rpcClient != nil {
			_, _ = m.PublishPrekeyBundle(bundle)
		}
		return bundle, nil
	}
	bundle, err := m.BuildPrekeyBundle(metadata, "", "")
	if err != nil {
		return PrekeyBundle{}, err
	}
	if m.rpcClient != nil {
		_, _ = m.PublishPrekeyBundle(bundle)
	}
	return bundle, nil
}

// VerifyRemotePrekeyBundle verifies a prekey bundle against a DID document.
func (m *PrekeyManager) VerifyRemotePrekeyBundle(bundle PrekeyBundle, didDocument map[string]any) error {
	return VerifyPrekeyBundle(bundle, didDocument)
}

// EnsureFreshOneTimePrekeys makes sure at least minCount OPKs are available locally.
func (m *PrekeyManager) EnsureFreshOneTimePrekeys(minCount int) ([]OneTimePrekey, error) {
	if m.oneTimePrekeyStore == nil || minCount <= 0 {
		return nil, nil
	}
	current, err := m.listOneTimePrekeys()
	if err != nil {
		return nil, err
	}
	if len(current) >= minCount {
		return current, nil
	}
	prefix := time.Now().UTC().UnixNano()
	for i := len(current); i < minCount; i++ {
		if _, _, err := m.GenerateOneTimePrekey(fmt.Sprintf("opk-%d-%03d", prefix, i)); err != nil {
			return nil, err
		}
	}
	return m.listOneTimePrekeys()
}

func (m *PrekeyManager) listOneTimePrekeys() ([]OneTimePrekey, error) {
	if m.oneTimePrekeyStore == nil {
		return nil, nil
	}
	result, err := m.oneTimePrekeyStore.ListOneTimePrekeys()
	if err != nil {
		return nil, err
	}
	sort.Slice(result, func(i, j int) bool {
		return result[i].KeyID < result[j].KeyID
	})
	return result, nil
}

func oneTimePrekeysToAny(values []OneTimePrekey) []any {
	result := make([]any, 0, len(values))
	for _, value := range values {
		result = append(result, map[string]any{
			"key_id":          value.KeyID,
			"public_key_b64u": value.PublicKeyB64U,
		})
	}
	return result
}
